### DISCUSSION TITLE
N-N relationship

### DISCUSSION THREAD

**Discussion by ryan-zheng-teki** [lecousin/lc-spring-data-r2dbc#2] [General]
Hi, Will you support N-N relationship?

**Comment by lecousin** [comment]
Hi, yes I will try to add it quickly

**Reply by ryan-zheng-teki** [reply]
Thank you. I am currently using the library you created to create applications :)

**Comment by lecousin** [comment]
Hi,
Support of n-n relationship has been added to version 0.4.0.

To use it the new @JoinTable annotation has been added. Currently a field with this annotation is retricted to be a Set.

A test is present in [the manytomanymodel test package](https://github.com/lecousin/lc-spring-data-r2dbc/tree/0.4.0/core/src/test/java/net/lecousin/reactive/data/relational/test/manytomanymodel) and may be used as example.

Please let me know if it answers to your needs.

**Reply by ryan-zheng-teki** [reply]
Hi, Thanks.  The many-to-many is working. But i see that there is no automatic scan of @Table annotation to generate the schema which is similar to JPA. Are you planning to add it as well?

**Reply by ryan-zheng-teki** [reply]
Hi, Can i work on the same project with you? I would like to add the functionality of automatic package scanning similar to Hibernate, and generate the schema during startup? So we don't have to manually configure the entity in xml file.

**Reply by lecousin** [reply]
Hi, yes of course any contribution is welcome :)
However this one may be not so simple:

The reason of the yaml file to declare the entities is because of the enhancement part during  initialization.
The enhancement allows to rewrite part of the entity classes (to add a state property, rewrite the setters, implement the lazy loading methods). But to be able to rewrite the classes, this step needs to be done before those classes are loaded into the JVM.

If we want to scan packages looking for classes with the @Table annotation, we will need to do it by analyzing the .class file, not reflection which needs the class to be loaded. Or to find another way to enhance classes...

May be we could use [this class](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/io/support/PathMatchingResourcePatternResolver.html) provided by Spring to scan classpath and find .class files in specific packages.
But probably we will need at least to declare those packages, because analyzing all .class files in the class path may slow down application startup... Or we may let the developer choose and document it: declare all antities, declare only packages, or declare nothing. Knowing that more you declare, faster the application will startup.

**Reply by ryan-zheng-teki** [reply]
I agree with you. In most enterprise applications, the models are declared normally in a central package. All the model classes reside inside that package. Then all the *.class found will be those entities declared with @Table annotation. I can work on this. :)

**Comment by lecousin** [comment]
Since version 0.5.1 the yaml file is now optional, but recommended for performance. If not present, the classpath is scanned to find entities automatically
